components/mpas-ocean/src/shared/mpas_ocn_vertical_remap.F

```diff
@@ -0,0 +1,436 @@
+         heightCellNew(minLevelCell(iCell), iCell) = heightCellNow(minLevelCell(iCell), iCell)
+         do k = maxLevelCell(iCell), minLevelCell(iCell)+1, -1
+            heightCellNew(k, iCell) = &
+               heightCellNew(k + 1, iCell) - layerThicknessNew(k, iCell)
+         end do
+      end do
+      !$omp end do
+      !$omp end parallel
+
+      call ocn_diagnostic_solve_layerThicknessEdge(normalVelocity, layerThickness, layerThickEdge)
+      call ocn_diagnostic_solve_layerThicknessEdge(normalVelocity, layerThicknessNew, layerThickEdgeNew)
+
+      !$omp parallel
+      !$omp do schedule(runtime) private(k,kTop,kBot,cell1,cell2)
+      do iEdge = 1, nEdges
+
+         cell1 = cellsOnEdge(1,iEdge)
+         cell2 = cellsOnEdge(2,iEdge)
+
+         kTop = minLevelEdgeBot(iEdge)
+         kBot = maxLevelEdgeTop(iEdge)
+
+         ! We only need to define the kBot+1 depth for one active cell
+         ! as the absolute kBot+1 depth has no effect on remapping
+         if (cell1 <= nCellsAll .and. maxLevelCell(cell1) > 0) then
+            heightEdgeNow(:, iEdge) = heightCellNow(kBot+1, cell1)
+            heightEdgeNew(:, iEdge) = heightCellNew(kBot+1, cell1)
+         elseif (cell2 <= nCellsAll .and. maxLevelCell(cell2) > 0) then
+            heightEdgeNow(:, iEdge) = heightCellNow(kBot+1, cell2)
+            heightEdgeNew(:, iEdge) = heightCellNew(kBot+1, cell2)
+         else
+            call mpas_log_write('No valid cell for edge remapping', MPAS_LOG_WARN)
+            heightEdgeNow(:,iEdge) = 0.0_RKIND
+            cycle
+         endif
+
+         do k = kBot, kTop, -1
+            heightEdgeNow(k, iEdge) = heightEdgeNow(k+1, iEdge) - layerThickEdge(k, iEdge)
+            heightEdgeNew(k, iEdge) = heightEdgeNew(k+1, iEdge) - layerThickEdgeNew(k, iEdge)
+         end do
+
+      end do
+      !$omp end do
+      !$omp end parallel
+
+      ! Assign new layerThicknesses
+      layerThickness = layerThicknessNew
+
+      !----------------------------------------------------------------------------
+      ! ACTUAL REMAPPING FROM HERE
+
+      nDoFs = 1 ! always 1 for MPAS, it'd be >1 for DG methods (with more DoF per layer)
+
+      ! how many variables to be remapped concurrently per column
+      if (associated(highFreqThickness) .and. associated(lowFreqDivergence)) then
+         nVars = 3
+      else
+         nVars = 1
+      endif
+
+      allocate(dataNow(nDoFs, nVars, nVertLevels+1))
+      allocate(dataNew(nDoFs, nVars, nVertLevels+1))
+      allocate(bcUpper(nVars))
+      allocate(bcLower(nVars))
+
+      ! set boundary conditions
+      bcUpper%bcopt = bc_upper
+      bcLower%bcopt = bc_lower
+
+      dataNow = 0.0_RKIND
+      dataNew = 0.0_RKIND
+
+      call work%init(nVertLevels + 1, nVars, opts) ! this is internal workspace for PPR
+
+      ! Remap all edge-centred variables from heightEdgeNow to heightEdgeNew
+      if (associated(highFreqThickness) .and. associated(lowFreqDivergence)) then
```

Comment:
This and the similar line above don't seem to be working as expected, at least in my gfortran 10.3.0 testing.  When nothing is assigned to these pointers, they still are returning `.true.` from `associated()`.  It seems like this is only useful if the target has been nullified.  I'm not sure I fully understand this.  But I think it would be better to use:
```suggestion
      if (config_use_freq_filtered_thickness) then
```